src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.Cancun.cs

```diff
@@ -16,8 +14,66 @@ public partial class EngineRpcModule : IEngineRpcModule
 {
     private readonly IAsyncHandler<byte[], GetPayloadV3Result?> _getPayloadHandlerV3;
 
-    public Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV3(ExecutionPayload executionPayload) =>
-        NewPayload(executionPayload, 3);
+    public Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV3(ExecutionPayload executionPayload, byte[][] blobVersionedHashes)
+    {
+        if (blobVersionedHashes is null)
+        {
+            string error = "Blob versioned hashes must be set";
+            if (_logger.IsWarn) _logger.Warn(error);
+            return ResultWrapper<PayloadStatusV1>.Fail(error, ErrorCodes.InvalidParams);
+        }
+
+        if (executionPayload is null)
+        {
+            string error = "Execution payload must be set";
+            if (_logger.IsWarn) _logger.Warn(error);
+            return ResultWrapper<PayloadStatusV1>.Fail(error, ErrorCodes.InvalidParams);
+        }
+
+        int index = 0;
+        foreach (Transaction tx in executionPayload.GetTransactions())
+        {
+            if (tx.BlobVersionedHashes is null)
+            {
+                continue;
+            }
+
+            foreach (byte[]? blobVersionedHash in tx.BlobVersionedHashes)
+            {
+                if (index == blobVersionedHashes.Length
+                    || blobVersionedHash is null
+                    || blobVersionedHashes[index] is null
+                    || !blobVersionedHash.SequenceEqual(blobVersionedHashes[index]))
+                {
+                    string error = "Blob versioned hashes do not match";
+                    if (_logger.IsWarn) _logger.Warn(error);
+                    return ResultWrapper<PayloadStatusV1>.Success(
+                        new PayloadStatusV1
+                        {
+                            Status = PayloadStatus.Invalid,
+                            LatestValidHash = null,
+                            ValidationError = error
+                        });
+                }
```

Comment:
Why do we need this check at all it seems redundant (I know it is in spec)?

Reply:
CL can't into RLP, but can calculate the hashes from the commitments that EL sends to CL in previous response. So it sends hashes to confirm that they are the same as in transactions' RLP. It's the reason we tried SSZ, so CL can decode transactions and check some stuff(like related to blobs).